### ISSUE TITLE
Fails to create file with 'mode=a' when non-ASCII in path

### ISSUE THREAD

**Issue by jnsebgosselin** [h5py/h5py#896]
To assist reproducing bugs, please include the following:
 * Operating System : Win7 64bit
 * Python version : 3.4.4 
 * Where Python was acquired : manual installation, downloaded from www.python.org
 * h5py version : 2.6.0
 * HDF5 version : 1.8.18
 * The full traceback/stack trace shown (if it appears):

When a non-ASCII character is present in the path, instantiating a File object fails when the file does not exists and when in  "Read/write if exists, create otherwise (default)" mode (mode=a).

The code below can be use to reproduce this behavior systematically on my system:
```
# -*- coding: utf-8 -*-

import h5py
import os
print(h5py.version.info)

path = 'testé'
if not os.path.exists(path):
    os.makedirs(path)

file = os.path.join('testé', 'test.h5')

print("h5py.File(file, mode='w') works whether the file exists or not")
h5py.File(file, mode='w')

print("h5py.File(file, mode='a') works if the file already exists")
h5py.File(file, mode='a')

print("h5py.File(file, mode='a') fails if the file does not already exist\n")
os.remove(file)
h5py.File(file, mode='a')
```
which results in the following traceback:

Summary of the h5py configuration
---------------------------------

h5py    2.6.0
HDF5    1.8.18
Python  3.4.4 (v3.4.4:737efcadf5a6, Dec 20 2015, 20:20:57) [MSC v.1600 64 bit (AMD64)]
sys.platform    win32
sys.maxsize     9223372036854775807
numpy   1.11.1

h5py.File(file, mode='w') works whether the file exists or not
h5py.File(file, mode='a') works if the file already exists
h5py.File(file, mode='a') fails if the file does not already exist

Traceback (most recent call last):
  File "C:\Users\jnsebgosselin\bug_nonASCII_path.py", line 26, in <module>
    h5py.File(file, mode='a')
  File "c:\python34\lib\site-packages\h5py\_hl\files.py", line 271, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, swmr=swmr)
  File "c:\python34\lib\site-packages\h5py\_hl\files.py", line 113, in make_fid
    fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)
  File "h5py\_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-git\h5py\_objects.c:2853)
  File "h5py\_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-git\h5py\_objects.c:2811)
  File "h5py\h5f.pyx", line 78, in h5py.h5f.open (D:\Build\h5py\h5py-git\h5py\h5f.c:2130)
  File "h5py\defs.pyx", line 621, in h5py.defs.H5Fopen (D:\Build\h5py\h5py-git\h5py\defs.c:6912)
  File "h5py\_errors.pyx", line 123, in h5py._errors.set_exception (D:\Build\h5py\h5py-git\h5py\_errors.c:1495)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 33: invalid continuation byte

**Comment by aragilar** [comment]
Problem is https://github.com/h5py/h5py/blob/master/h5py/_errors.pyx#L123, which assumes we can decode filenames as utf8. This means that there is some error that we're getting from HDF5, but not decoding it correctly.

**Comment by tacaswell** [comment]
This may be windows specific bug, it works correctly on linux.

**Comment by jnsebgosselin** [comment]
It is probably due to the fact that unlike Linux, Windows does not use utf8 as default encoding of the OS.

**Comment by tacaswell** [comment]
I think this will be fixed in 3.6 (see https://www.python.org/dev/peps/pep-0529/) ?

If not, we should use the same `fsencode` logic we use in `_hl.compat` here, might be worth pushing that out of `_hl` so that we avoid circular imports or just vendor it a second time in `_errors.pyx`.

**Comment by jnsebgosselin** [comment]
Nice read, thanks @tacaswell.

I've recently updated my system since some clever people found a way to make PySide 1 works with Python 3.6. The error is still there, but maybe it is because the proposed change has not been implemented yet?

Below is the traceback of the bug with my new setup:

Summary of the h5py configuration
---------------------------------

h5py    2.7.0.post0
HDF5    1.10.1
Python  3.6.1 (v3.6.1:69c0db5, Mar 21 2017, 18:41:36) [MSC v.1900 64 bit (AMD64)]
sys.platform    win32
sys.maxsize     9223372036854775807
numpy   1.13.0

h5py.File(file, mode='w') works whether the file exists or not
h5py.File(file, mode='a') works if the file already exists
h5py.File(file, mode='a') fails if the file does not already exist

Traceback (most recent call last):
  File "C:\Users\jsgosselin\untitled1.py", line 21, in <module>
    h5py.File(file, mode='a')
  File "c:\python36\lib\site-packages\h5py\_hl\files.py", line 269, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, swmr=swmr)
  File "c:\python36\lib\site-packages\h5py\_hl\files.py", line 111, in make_fid
    fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)
  File "h5py\_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-hdf5110-git\h5py\_objects.c:2856)
  File "h5py\_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-hdf5110-git\h5py\_objects.c:2814)
  File "h5py\h5f.pyx", line 78, in h5py.h5f.open (D:\Build\h5py\h5py-hdf5110-git\h5py\h5f.c:2141)
  File "h5py\defs.pyx", line 621, in h5py.defs.H5Fopen (D:\Build\h5py\h5py-hdf5110-git\h5py\defs.c:7139)
  File "h5py\_errors.pyx", line 123, in h5py._errors.set_exception (D:\Build\h5py\h5py-hdf5110-git\h5py\_errors.c:1498)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 33: invalid continuation byte

**Comment by tacaswell** [comment]
My reading of that pep is the changes went in for 3.6.  I think the issue here is that we are getting bytes back from hdf5 (and going around all of the python / stdlib related things).  I think the only fix for this is the `fsencode` route.

**Comment by takluyver** [comment]
It looks like this was fixed in #1025, which went into 2.8.
